### PR TITLE
feat: space user profile page

### PR DESCRIPTION
**Pull request by wa0x6e**
### Summary

This PR add a new user profile page, under each space, summarizing the user activities for the given space

In order to facilitate reviewing, this feature will be split into multiple PR, and this one will only handle the skeleton and routes to all pages, as well as the everything related to the `STATEMENT` tab. All other pages (delegators, proposals and votes) will be empty, and populating them will come in a future PR.

### How to test

1. Go to a space leaderboard
2. Click on an user
3. It should redirect you to the space user profile page
4. For onchain spaces, it should show the correct number of proposals and votes
5. It should display the user's voting power on this space
6. It should show the user's statement if available
7. When it's your own profile, you should be able to edit it

### TODO 

- [x] Display voting power
- [x] Edit statement
- [x] Add new fields (`status`, `discourse`) to SDK - https://github.com/snapshot-labs/snapshot.js/pull/1023
- [x] Add new fields (`status`, `discourse`) to Sequencer - https://github.com/snapshot-labs/snapshot-sequencer/pull/386
- [x] Add new fields (`status`, `discourse`) to Hub - https://github.com/snapshot-labs/snapshot-hub/pull/870

### CODE REVIEW

File: apps/ui/src/views/SpaceUser.vue [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+  }
+
+  return value;
+});
+
+const navigation = computed(() => [
+  { label: 'Statement', route: 'space-user-statement' },
+  { label: 'Delegators', route: 'space-user-delegators' },
+  { label: 'Proposals', route: 'space-user-proposals', count: userActivity.value?.proposal_count },
+  { label: 'Votes', route: 'space-user-votes', count: userActivity.value?.vote_count }
```
Comment (ChaituVR) [inline apps/ui/src/views/SpaceUser.vue:58]: On design, it is "LATEST VOTES"
Untitled 6

File: packages/sx.js/src/clients/offchain/ethereum-sig/types.ts [hunk 2]
```diff
@@ -134,3 +134,16 @@ export const updateUserTypes = {
     { name: 'profile', type: 'string' }
   ]
 };
+
+export const updateStatementTypes = {
+  Statement: [
+    { name: 'from', type: 'address' },
```
Comment (ChaituVR) [inline packages/sx.js/src/clients/offchain/ethereum-sig/types.ts:140]: `from` is now `string` right? 🤔

File: packages/sx.js/src/clients/offchain/ethereum-sig/index.ts [hunk 3]
```diff
@@ -85,9 +89,8 @@ export class EthereumSig {
       timestamp: parseInt((Date.now() / 1e3).toFixed()),
       ...message
     };
-
     const signature = await signer._signTypedData(domain, types, EIP712Message);
-
+    console.log(signature);
```
Comment (ChaituVR) [inline packages/sx.js/src/clients/offchain/ethereum-sig/index.ts:93]:
```suggestion
```

File: apps/ui/src/views/SpaceUser.vue [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  if (props.space.voting_power_symbol) {
+    return `${value} ${props.space.voting_power_symbol}`;
+  }
+
+  return value;
+});
+
+const navigation = computed(() => [
+  { label: 'Statement', route: 'space-user-statement' },
+  { label: 'Delegators', route: 'space-user-delegators' },
```
Comment (ChaituVR) [inline apps/ui/src/views/SpaceUser.vue:56]: On design, we show could on delegators tab also 

image

File: apps/ui/src/views/SpaceUser/Votes.vue [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<script setup lang="ts"></script>
+
+<template>
+  <div></div>
+</template>
```
Comment (ChaituVR) [inline apps/ui/src/views/SpaceUser/Votes.vue:5]: No design for delegates, votes and proposals? right now it show empty
Untitled 7

**Review by ChaituVR — CHANGES_REQUESTED** [review]
looks like typecheck is failing

**Comment by Sekhmet** [comment]
@wa0x6e removed `Closes:` message as this PR is part of PR chain, after we merge this one it will still miss non-statement tabs.

File: packages/sx.js/src/clients/offchain/ethereum-sig/types.ts (hunk 2 above)
Comment (wa0x6e) [reply packages/sx.js/src/clients/offchain/ethereum-sig/types.ts:140]: Not until the starknet PR is merged

File: apps/ui/src/views/SpaceUser/Votes.vue (hunk 5 above)
Comment (wa0x6e) [reply apps/ui/src/views/SpaceUser/Votes.vue:5]: Indeed. And UI is also not ready for these pages, afaik

File: apps/ui/src/views/SpaceUser.vue (hunk 4 above)
Comment (Sekhmet) [reply apps/ui/src/views/SpaceUser.vue:56]: I believe it's about missing "count" next to Delegators (on UI we have counter that shows 6, but here we don't show any counter).

Comment (ChaituVR) [reply apps/ui/src/views/SpaceUser.vue:56]: oops yes, sorry for the typos

**Comment by wa0x6e** [comment]
Screenshot 2024-07-17 at 17 05 42

@bonustrack  This page need some more UI when no statement exist, else it will look like the page fail to load completely

**Comment by aurelianoB** [comment]
> this page need some more UI when no statement exist, else it will look like the page fail to load completely

The design for no statement already exists here: https://www.figma.com/design/hmHTau1Y1KYzrBBKLQr5QV/%23design?node-id=5500-6986&t=cEXvZzxGWmeuUUT8-4

File: apps/ui/src/views/SpaceUser/Delegators.vue [hunk 6]
```diff
@@ -1,5 +1,35 @@
...
+  user: User;
+}>();
+</script>
 
 <template>
-  <div></div>
+  <div
+    class="bg-skin-bg border-b sticky top-[112px] lg:top-[113px] z-40 flex w-full font-medium space-x-1"
+  >
+    <div class="pl-4 w-[60%] flex truncate">Delegation contract</div>
```
Comment (aurelianoB) [inline apps/ui/src/views/SpaceUser/Delegators.vue:17]: What's the purpose of this?

Comment (wa0x6e) [reply apps/ui/src/views/SpaceUser/Delegators.vue:17]: Showing the related delegation contract, since you can delegate on multiple contract.

Comment (aurelianoB) [reply apps/ui/src/views/SpaceUser/Delegators.vue:17]: I may be missing some context on some discussion you had elsewhere, but judging on the design (https://www.figma.com/design/hmHTau1Y1KYzrBBKLQr5QV/%23design?node-id=5436-5066&t=YDxtUhTDuXjJRPkm-4) the idea is to have a list of individual delegators here, not an aggregation of total delegators/delegatedVP per network/contract.

That said it's true that the design is lacking a solution for when there are multiple delegation contracts in place.

**Comment by wa0x6e** [comment]
I don’t see the delegators pages in the figma. It’s the only one missing

**Comment by wa0x6e** [comment]
My bad, the tab link was not colored correctly, the page exists.

File: apps/ui/src/views/SpaceUser.vue [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+import { addressValidator as isValidAddress } from '@/helpers/validation';
+import {
+  _vp,
+  autoLinkText,
+  getCacheHash,
+  getSocialNetworksLink,
+  shortenAddress
+} from '@/helpers/utils';
+import { getNetwork, supportsNullCurrent } from '@/networks';
+import type { Space, SpaceMetadataDelegation, UserActivity } from '@/types';
```
Comment (ChaituVR) [inline apps/ui/src/views/SpaceUser.vue:11]:
```suggestion
import type { Space, UserActivity } from '@/types';
```

File: apps/ui/src/views/SpaceUser.vue [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+import type { VotingPower, VotingPowerStatus } from '@/networks/types';
+import type { Delegate } from '@/composables/useDelegates';
```
Comment (ChaituVR) [inline apps/ui/src/views/SpaceUser.vue:13]:
```suggestion
```

File: apps/ui/src/views/SpaceUser.vue [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    props.space.id,
+    {
+      limit: 1,
+      skip: 0
+    },
+    'vote_count-desc',
+    userId.value
+  );
+
+  if (users[0]) userActivity.value = users[0];
```
Comment (ChaituVR) [inline apps/ui/src/views/SpaceUser.vue:75]: When i go here http://localhost:8080/#/s:aave.eth/profile/0xF1dF824419879Bb8a7E758173523F88EfB7Af193

I see `0` proposals and `565` votes
Untitled 2

On DB I see:
Untitled 4

On the request input I don't see user id:
Untitled 5

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: apps/ui/src/views/Proposal/Votes.vue [hunk 10]
```diff
@@ -163,9 +163,10 @@ watch([sortBy, choiceFilter], () => {
           <UiStamp :id="vote.voter.id" :size="32" />
           <router-link
             :to="{
-              name: 'user',
+              name: 'space-user-statement',
               params: {
-                id: vote.voter.id
+                id: `${proposal.network}:${proposal.space.id}`,
+                user: vote.voter.id
```
Comment (ChaituVR) [inline apps/ui/src/views/Proposal/Votes.vue:169]: Similars to votes page on proposal, i think we should also redirect votes modal to `space-user-statement`, right now It redirect to `user`
Untitled 6

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: apps/ui/src/views/SpaceUser.vue (hunk 9 above)
Comment (wa0x6e) [reply apps/ui/src/views/SpaceUser.vue:75]: Fixed, the offchain leaderboard PR was not ready when this PR was submitted

File: apps/ui/src/views/SpaceUser.vue [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+  [resolved, networkId, address],
+  async ([resolved, networkId, address]) => {
+    if (!resolved || !networkId || !address) return;
+
+    spacesStore.fetchSpace(address, networkId);
+  },
+  {
+    immediate: true
+  }
+);
```
Comment (ChaituVR) [inline apps/ui/src/views/SpaceUser.vue:158]: This is causing duplicate `space` query

File: apps/ui/src/views/SpaceUser.vue [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+    if (isValidAddress(id)) {
+      await usersStore.fetchUser(id);
+      await loadUserActivity();
+      loadDelegates();
+      getVotingPower();
+    }
+
+    loaded.value = true;
+  },
+  { immediate: true }
```
Comment (ChaituVR) [inline apps/ui/src/views/SpaceUser.vue:145]: If i go here http://localhost:8080/#/sn:0x039a5539bf1866da915779bb6270d1b839ca82cb0acd4eefbc17ba99ebc3a021/profile/0x013477227a72edc53a237231ec9f4c7180c93386664b1e3a1b5d181c354e5f39

Replace user address with `0x03c4dd9305d6b3a2c1ba7ed06c2285125751600c821ec45279e3760db341d67e` in URL, who doesn't any votes or proposals on this proposal
It shows previous values 🙈 when i reload it works fine

**Review by ChaituVR — APPROVED** [review]
Apart from above two non-blocking issues, everything looks good to me

tAck, should we wait for @Sekhmet ?

File: apps/ui/src/components/EditorStatement.vue [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    emit('close');
+  } finally {
+    sending.value = false;
+  }
+}
+
+watchEffect(async () => {
+  formValidated.value = false;
+
+  formErrors.value = await formValidator.validateAsync({
```
Comment (Sekhmet) [inline apps/ui/src/components/EditorStatement.vue:63]: We could use non-async version there as we don't use async properties (`ens-or-address`).

File: apps/ui/src/views/SpaceUser.vue [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+
+  return value;
+});
+
+const navigation = computed(() => [
+  { label: 'Statement', route: 'space-user-statement' },
+  { label: 'Delegators', route: 'space-user-delegators', count: delegatesCount.value },
+  { label: 'Proposals', route: 'space-user-proposals', count: userActivity.value?.proposal_count },
+  { label: 'Latest votes', route: 'space-user-votes', count: userActivity.value?.vote_count }
+]);
```
Comment (Sekhmet) [inline apps/ui/src/views/SpaceUser.vue:60]:
```suggestion
const navigation = computed(() => [
  { label: 'Statement', route: 'space-user-statement' },
  // { label: 'Delegators', route: 'space-user-delegators', count: delegatesCount.value },
  // { label: 'Proposals', route: 'space-user-proposals', count: userActivity.value?.proposal_count },
  // { label: 'Latest votes', route: 'space-user-votes', count: userActivity.value?.vote_count }
]);
```

I think we should comment out other tabs if we want to merge it in steps, as those other pages will look like something is broken as those sections are empty (and we have some real users using the app already).

**Review by Sekhmet — COMMENTED** [review]
Other than that it looks good to me, great job!

**Comment by wa0x6e** [comment]
Maybe wait for https://github.com/snapshot-labs/sx-monorepo/pull/520 so we can add validation to the statement body

**Review by Sekhmet — APPROVED** [review]
tACK

**Comment by Sekhmet** [comment]
> Maybe wait for https://github.com/snapshot-labs/sx-monorepo/pull/520 so we can add validation to the statement body

We can if you want, but we could merge this PR and add support for error state in statement in #520. I'd imagine it might take some time to get this PR to be in ready state, getting design right etc.

**Comment by wa0x6e** [comment]
Merge conflict fixed, let's merge this one first

**Pull request merged**